File: mRemoteV1/Forms/ExportForm.vb
Imports System.ComponentModel
Imports mRemote3G.App
Imports mRemote3G.Config.Connections

Namespace Forms
    Public Class ExportForm
        Inherits Form

#Region "Public Properties"

        Public Property FileName As String
            Get
                Return txtFileName.Text
            End Get
            Set
                txtFileName.Text = value
            End Set
        End Property

        Public Property SaveFormat As ConnectionsSave.Format
            Get
                Dim exportFormat = TryCast(cboFileFormat.SelectedItem, ExportFormat)
                If exportFormat Is Nothing Then
                    Return ConnectionsSave.Format.mRXML
                Else
                    Return exportFormat.Format
                End If
            End Get
            Set
                For Each item As Object In cboFileFormat.Items
                    Dim exportFormat = TryCast(item, ExportFormat)
                    If exportFormat Is Nothing Then Continue For
                    If exportFormat.Format = value Then
                        cboFileFormat.SelectedItem = item
                        Exit For
                    End If
                Next
            End Set
        End Property

        Public Property Scope As ExportScope
            Get
                If rdoExportSelectedFolder.Checked Then
                    Return ExportScope.SelectedFolder
                ElseIf rdoExportSelectedConnection.Checked Then
                    Return ExportScope.SelectedConnection
                Else
                    Return ExportScope.Everything
                End If
            End Get
            Set
                Select Case value
                    Case ExportScope.Everything
                        rdoExportEverything.Checked = True
                    Case ExportScope.SelectedFolder
                        rdoExportSelectedFolder.Checked = True
                    Case ExportScope.SelectedConnection
                        rdoExportSelectedConnection.Checked = True
                End Select
            End Set
        End Property

        Private _selectedFolder As TreeNode

        Public Property SelectedFolder As TreeNode
            Get
                Return _selectedFolder
            End Get
            Set
                _selectedFolder = value
                If value Is Nothing Then
                    lblSelectedFolder.Text = String.Empty
                Else
                    lblSelectedFolder.Text = value.Text
                End If
                rdoExportSelectedFolder.Enabled = (value IsNot Nothing)
            End Set
        End Property

        Private _selectedConnection As TreeNode

        Public Property SelectedConnection As TreeNode
            Get
                Return _selectedConnection
            End Get
            Set
                _selectedConnection = value
                If value Is Nothing Then
                    lblSelectedConnection.Text = String.Empty
                Else
                    lblSelectedConnection.Text = value.Text
                End If
                rdoExportSelectedConnection.Enabled = (value IsNot Nothing)
            End Set
        End Property

        Public Property IncludeUsername As Boolean
            Get
                Return chkUsername.Checked
            End Get
            Set
                chkUsername.Checked = value
            End Set
        End Property

        Public Property IncludePassword As Boolean
            Get
                Return chkPassword.Checked
            End Get
            Set
                chkPassword.Checked = value
            End Set
        End Property

        Public Property IncludeDomain As Boolean
            Get
                Return chkDomain.Checked
            End Get
            Set
                chkDomain.Checked = value
            End Set
        End Property

        Public Property IncludeInheritance As Boolean
            Get
                Return chkInheritance.Checked
            End Get
            Set
                chkInheritance.Checked = value
            End Set
        End Property

#End Region

#Region "Constructors"

        Public Sub New()
            InitializeComponent()

            Runtime.FontOverride(Me)

            SelectedFolder = Nothing
            SelectedConnection = Nothing

            btnOK.Enabled = False
        End Sub

#End Region

#Region "Private Methods"

#Region "Event Handlers"

        Private Sub ExportForm_Load(sender As Object, e As EventArgs) Handles Me.Load
            cboFileFormat.Items.Clear()
            cboFileFormat.Items.Add(New ExportFormat(ConnectionsSave.Format.mRXML))
            cboFileFormat.Items.Add(New ExportFormat(ConnectionsSave.Format.mRCSV))
            cboFileFormat.Items.Add(New ExportFormat(ConnectionsSave.Format.vRDCSV))
            cboFileFormat.SelectedIndex = 0

            ApplyLanguage()
        End Sub

        Private Sub txtFileName_TextChanged(sender As Object, e As EventArgs) Handles txtFileName.TextChanged
            btnOK.Enabled = Not String.IsNullOrEmpty(txtFileName.Text)
        End Sub

        Private Sub btnBrowse_Click(sender As Object, e As EventArgs) Handles btnBrowse.Click
            Using saveFileDialog As New SaveFileDialog()
                With saveFileDialog
                    .CheckPathExists = True
                    .InitialDirectory = Environment.GetFolderPath(Environment.SpecialFolder.Personal)
                    .OverwritePrompt = True

                    Dim fileTypes As New List(Of String)
                    fileTypes.AddRange({Language.Language.strFiltermRemoteXML, "*.xml"})
                    fileTypes.AddRange({Language.Language.strFiltermRemoteCSV, "*.csv"})
                    fileTypes.AddRange({Language.Language.strFiltervRD2008CSV, "*.csv"})
                    fileTypes.AddRange({Language.Language.strFilterAll, "*.*"})

                    .Filter = String.Join("|", fileTypes.ToArray())
                End With

                If Not saveFileDialog.ShowDialog(Me) = DialogResult.OK Then Return

                txtFileName.Text = saveFileDialog.FileName
            End Using
        End Sub

        Private Sub btnOK_Click(sender As Object, e As EventArgs) Handles btnOK.Click
            DialogResult = DialogResult.OK
        End Sub

        Private Sub btnCancel_Click(sender As Object, e As EventArgs) Handles btnCancel.Click
            DialogResult = DialogResult.Cancel
        End Sub

#End Region

        Private Sub ApplyLanguage()
            Text = Language.Language.strExport

            grpFile.Text = Language.Language.strExportFile
            lblFileName.Text = Language.Language.strLabelFilename
            btnBrowse.Text = Language.Language.strButtonBrowse
            lblFileFormat.Text = Language.Language.strFileFormatLabel

            grpItems.Text = Language.Language.strExportItems
            rdoExportEverything.Text = Language.Language.strExportEverything
            rdoExportSelectedFolder.Text = Language.Language.strExportSelectedFolder
            rdoExportSelectedConnection.Text = Language.Language.strExportSelectedConnection

            grpProperties.Text = Language.Language.strExportProperties
            chkUsername.Text = Language.Language.strCheckboxUsername
            chkPassword.Text = Language.Language.strCheckboxPassword
            chkDomain.Text = Language.Language.strCheckboxDomain
            chkInheritance.Text = Language.Language.strCheckboxInheritance
            lblUncheckProperties.Text = Language.Language.strUncheckProperties

            btnOK.Text = Language.Language.strButtonOK
            btnCancel.Text = Language.Language.strButtonCancel
        End Sub

#End Region

#Region "Public Enumerations"

        Public Enum ExportScope
            Everything
            SelectedFolder
            SelectedConnection
        End Enum

#End Region

#Region "Private Classes"

        <ImmutableObject(True)>
        Private Class ExportFormat

#Region "Public Properties"

            Private ReadOnly _format As ConnectionsSave.Format

            Public ReadOnly Property Format As ConnectionsSave.Format
                Get
                    Return _format
                End Get
            End Property

#End Region

#Region "Constructors"

            Public Sub New(format As ConnectionsSave.Format)
                _format = format
            End Sub

#End Region

#Region "Public Methods"

            Public Overrides Function ToString() As String
                Select Case Format
                    Case ConnectionsSave.Format.mRXML
                        Return Language.Language.strMremoteNgXml
                    Case ConnectionsSave.Format.mRCSV
                        Return Language.Language.strMremoteNgCsv
                    Case ConnectionsSave.Format.vRDCSV
                        Return Language.Language.strVisionAppRemoteDesktopCsv
                    Case Else
                        Return Format.ToString()
                End Select
            End Function

#End Region
        End Class

#End Region
    End Class
End Namespace

File: mRemoteV1/CS/UI/Forms/ExportForm.cs
using System.Collections.Generic;
using System;
using AxWFICALib;
using System.Drawing;
using System.Diagnostics;
using System.Data;
using AxMSTSCLib;
using Microsoft.VisualBasic;
using System.Collections;
using System.Windows.Forms;
using System.ComponentModel;
using mRemoteNG.Config;
using mRemoteNG.App;
using mRemoteNG.My;


namespace mRemoteNG.Forms
{
    public partial class ExportForm : Form
	{
        #region Public Properties
        public string FileName
		{
			get
			{
				return txtFileName.Text;
			}
			set
			{
				txtFileName.Text = value;
			}
		}
			
        public Config.Connections.Save.Format SaveFormat
		{
			get
			{
				ExportFormat exportFormat = cboFileFormat.SelectedItem as ExportFormat;
				if (exportFormat == null)
				{
					return Config.Connections.Save.Format.mRXML;
				}
				else
				{
					return exportFormat.Format;
				}
			}
			set
			{
				foreach (object item in cboFileFormat.Items)
				{
					ExportFormat exportFormat = item as ExportFormat;
					if (exportFormat == null)
					{
						continue;
					}
					if (exportFormat.Format == value)
					{
						cboFileFormat.SelectedItem = item;
						break;
					}
				}
			}
		}
			
        public ExportScope Scope
		{
			get
			{
				if (rdoExportSelectedFolder.Checked)
				{
					return ExportScope.SelectedFolder;
				}
				else if (rdoExportSelectedConnection.Checked)
				{
					return ExportScope.SelectedConnection;
				}
				else
				{
					return ExportScope.Everything;
				}
			}
			set
			{
				switch (value)
				{
					case ExportScope.Everything:
						rdoExportEverything.Checked = true;
						break;
					case ExportScope.SelectedFolder:
						rdoExportSelectedFolder.Checked = true;
						break;
					case ExportScope.SelectedConnection:
						rdoExportSelectedConnection.Checked = true;
						break;
				}
			}
		}
			
		private TreeNode _selectedFolder;
        public TreeNode SelectedFolder
		{
			get
			{
				return _selectedFolder;
			}
			set
			{
				_selectedFolder = value;
				if (value == null)
				{
					lblSelectedFolder.Text = string.Empty;
				}
				else
				{
					lblSelectedFolder.Text = value.Text;
				}
				rdoExportSelectedFolder.Enabled = value != null;
			}
		}
			
		private TreeNode _selectedConnection;
        public TreeNode SelectedConnection
		{
			get
			{
				return _selectedConnection;
			}
			set
			{
				_selectedConnection = value;
				if (value == null)
				{
					lblSelectedConnection.Text = string.Empty;
				}
				else
				{
					lblSelectedConnection.Text = value.Text;
				}
				rdoExportSelectedConnection.Enabled = value != null;
			}
		}
			
        public bool IncludeUsername
		{
			get
			{
				return chkUsername.Checked;
			}
			set
			{
				chkUsername.Checked = value;
			}
		}
			
        public bool IncludePassword
		{
			get
			{
				return chkPassword.Checked;
			}
			set
			{
				chkPassword.Checked = value;
			}
		}
			
        public bool IncludeDomain
		{
			get
			{
				return chkDomain.Checked;
			}
			set
			{
				chkDomain.Checked = value;
			}
		}
			
        public bool IncludeInheritance
		{
			get
			{
				return chkInheritance.Checked;
			}
			set
			{
				chkInheritance.Checked = value;
			}
		}
        #endregion
			
        #region Constructors
		public ExportForm()
		{
			InitializeComponent();
				
			Runtime.FontOverride(this);
				
			SelectedFolder = null;
			SelectedConnection = null;
				
			btnOK.Enabled = false;
		}
        #endregion
			
        #region Private Methods
        #region Event Handlers
		public void ExportForm_Load(object sender, EventArgs e)
		{
			cboFileFormat.Items.Clear();
            cboFileFormat.Items.Add(new ExportFormat(Config.Connections.Save.Format.mRXML));
            cboFileFormat.Items.Add(new ExportFormat(Config.Connections.Save.Format.mRCSV));
            cboFileFormat.Items.Add(new ExportFormat(Config.Connections.Save.Format.vRDCSV));
			cboFileFormat.SelectedIndex = 0;
				
			ApplyLanguage();
		}
			
		public void txtFileName_TextChanged(System.Object sender, EventArgs e)
		{
			btnOK.Enabled = !string.IsNullOrEmpty(txtFileName.Text);
		}
			
		public void btnBrowse_Click(System.Object sender, EventArgs e)
		{
			using (SaveFileDialog saveFileDialog = new SaveFileDialog())
			{
				saveFileDialog.CheckPathExists = true;
				saveFileDialog.InitialDirectory = Environment.GetFolderPath(Environment.SpecialFolder.Personal);
				saveFileDialog.OverwritePrompt = true;
					
				List<string> fileTypes = new List<string>();
				fileTypes.AddRange(new[] {Language.strFiltermRemoteXML, "*.xml"});
				fileTypes.AddRange(new[] {Language.strFiltermRemoteCSV, "*.csv"});
				fileTypes.AddRange(new[] {Language.strFiltervRD2008CSV, "*.csv"});
				fileTypes.AddRange(new[] {Language.strFilterAll, "*.*"});
					
				saveFileDialog.Filter = string.Join("|", fileTypes.ToArray());
					
				if (!(saveFileDialog.ShowDialog(this) == DialogResult.OK))
				{
					return ;
				}
					
				txtFileName.Text = saveFileDialog.FileName;
			}
				
		}
			
		public void btnOK_Click(System.Object sender, EventArgs e)
		{
			DialogResult = DialogResult.OK;
		}
			
		public void btnCancel_Click(System.Object sender, EventArgs e)
		{
			DialogResult = DialogResult.Cancel;
		}
        #endregion
			
		private void ApplyLanguage()
		{
			Text = Language.strExport;
				
			grpFile.Text = Language.strExportFile;
			lblFileName.Text = Language.strLabelFilename;
			btnBrowse.Text = Language.strButtonBrowse;
			lblFileFormat.Text = Language.strFileFormatLabel;
				
			grpItems.Text = Language.strExportItems;
			rdoExportEverything.Text = Language.strExportEverything;
			rdoExportSelectedFolder.Text = Language.strExportSelectedFolder;
			rdoExportSelectedConnection.Text = Language.strExportSelectedConnection;
				
			grpProperties.Text = Language.strExportProperties;
			chkUsername.Text = Language.strCheckboxUsername;
			chkPassword.Text = Language.strCheckboxPassword;
			chkDomain.Text = Language.strCheckboxDomain;
			chkInheritance.Text = Language.strCheckboxInheritance;
			lblUncheckProperties.Text = Language.strUncheckProperties;
				
			btnOK.Text = Language.strButtonOK;
			btnCancel.Text = Language.strButtonCancel;
		}
        #endregion
			
        #region Public Enumerations
		public enum ExportScope
		{
			Everything,
			SelectedFolder,
			SelectedConnection
		}
        #endregion
			
        #region Private Classes
		[ImmutableObject(true)]
        private class ExportFormat
		{
            #region Public Properties
			private Config.Connections.Save.Format _format;
            public Config.Connections.Save.Format Format
			{
				get
				{
					return _format;
				}
			}
            #endregion
				
            #region Constructors
			public ExportFormat(Config.Connections.Save.Format format)
			{
				_format = format;
			}
            #endregion
				
            #region Public Methods
			public override string ToString()
			{
				switch (Format)
				{
					case Config.Connections.Save.Format.mRXML:
						return Language.strMremoteNgXml;
                    case Config.Connections.Save.Format.mRCSV:
						return Language.strMremoteNgCsv;
                    case Config.Connections.Save.Format.vRDCSV:
						return Language.strVisionAppRemoteDesktopCsv;
					default:
						return Format.ToString();
				}
			}
            #endregion
		}
        #endregion
	}
}
